### DISCUSSION TITLE
Extract frames pixel data

### DISCUSSION THREAD

**Discussion by Saveriu** [dcm4che/dcm4che#1442] [Q&A]
Hi, 
I try to create a Dicom Web Server with Spring boot and I'm currently implementing the frames resource:
/studies/{studyInstanceUID}/series/{seriesInstanceUID}/instances/{sopInstanceUID}/frames/{frames}.
I'm looking for a good snippet code to get multiple frames from a DicomInputStream.

One example of some implementations that I'm trying, not sure if the ImageInputStream is the correct class to return (maybe a byte[] is more accurate).

`public List<ImageInputStream> getFramePixelData(int[] frameIndexes) {
		var reader = (DicomImageReader) ImageIO.getImageReadersByFormatName("DICOM").next();
		reader.setInput(dicomInputStream);
		return Arrays.stream(frameIndexes)
				.mapToObj(frameIndex -> {
					try {
						return reader.iisOfFrame(frameIndex);
					} catch (IOException e) {
						throw new UncheckedIOException(e);
					}
				})
				.toList();		 
	}`
Do you have some piece of code that I can see or use ? Thanks!

**Comment by yaourt** [comment]
I'm having a similar need ...
I'm quite interested in any hints 🙏

**Comment by Saveriu** [comment]
I dig into the source code of dcm4che (DicomImageReader, EncapsulatedPixelDataImageInputStream, etc.) and I think my code won't work because when we call iisOfFrame we move the DicomInputStream cursor. 
But I can't even make it work with an enhanced.

```
void test() {
		try (
				var src = Files.newInputStream(ENHANCED);
				var dis = new DicomInputStream(src);
		) {
			var dicomImageReader =  (DicomImageReader) ImageIO.getImageReadersByFormatName("DICOM").next();
			dicomImageReader.setInput(dis);
			//to force the reading the internal call of readMetadata()
			dicomImageReader.getStreamMetadata();
			
			//return null ?!
			var iis = dicomImageReader.iisOfFrame(0);

			try (var dest = Files.newOutputStream(Path.of("frame_" + 0 + ".dat"))) {
				var buffer = new byte[1024];
                                //NPE
				while(iis.read(buffer) != -1) {
					dest.write(buffer);
				}
			}

		}
	}
```

**Reply by Saveriu** [reply]
Does the iisOfFrame method only work with "Fragments" pixel data?
